feat/FECA-37 Filter

YassenAhmed-gif opened the pull request:
This commit introduces product filtering capabilities to the categories feature. Users can now sort products by various criteria like price, newness, and discount.

Key changes:
- Added `createdAt` field to `ProductEntity` and `ProductDtoMapper` to support sorting by newness.
- Implemented `discountPercent` getter in `ProductEntity` for consistent discount calculation.
- Introduced `FilterType` enum and a generic `Filter` class in `core/Functions/filter.dart` to handle sorting logic.
- Updated `CategoriesState` to include `filteredProducts` and `currentFilter`, and added `displayProducts` getter to show either filtered or all products.
- Modified `CategoriesViewModel` to:
    - Include `applyFilter` and `clearFilter` methods.
    - Initialize `filteredProducts` when fetching products.
- Added `FilterBottomSheet` widget to allow users to select filter options.
- Integrated `FilterButton` with `FilterBottomSheet` in `CategoryScreen`.
- Updated `CategoryScreenBody` to display `state.displayProducts` and show a "No products found" message when applicable.
- Modified `ProductGrid` to use the `discountPercent` from `ProductEntity`.
- Added new translation keys for filter options (e.g., "sort_by", "lowest_price", "new_product") in English and Arabic.
- Fixed a typo in an import statement in `main.dart` (`ececute_navigation` to `execute_navigation`).

<!--
  🛠️ For Work In Progress, please use Draft PRs:
  https://github.blog/2019-02-14-introducing-draft-pull-requests/

  ✅ Before submitting:
  - Follow your project's contributing guide
  - Ensure code compiles and passes all tests
  - Use descriptive commit messages
  - Keep PRs small and focused
-->

## 📌 What type of PR is this? (check all applicable)


- [ ✅] ✨ Feature

- [ ] 🐛 Bug Fix
- [ ] 🧹 Refactor
- [ ] 🚀 Performance Optimization
- [ ] 🧾 Documentation Update
- [ ] ♿ Accessibility Fix
- [ ] 🔧 CI/CD or Config Change

---

## 📝 Description


_Describe the goal of this PR and what was changed._


---

## 🧩 Related Issues / Tickets


- Closes #
- Fixes #
- Related to #


---

## 🧪 QA Instructions

_How to test the changes manually?_

- [ ] Run the app on Android
- [ ] Run the app on iOS

- [ ] Check specific feature/page:
  - [ ] ...
- [ ] Verify responsiveness on different screen sizes


---

## 📸 Screenshots or Recordings (if UI changes)

https://github.com/user-attachments/assets/25770ff5-ae89-457b-834e-480c3f25ae87


---

## ♿ UI Accessibility Checklist (if applicable)

- [ ] Semantic widget structure used (`Semantics`, `Accessible`, etc.)
- [ ] Supports screen readers
- [ ] Good color contrast
- [ ] Tested with TalkBack/VoiceOver

---

## 🧪 Tests


- [ ❌] Unit tests added
- [❌] Widget tests added

- [ ] Integration tests added
- [ ] No tests needed (explain why):

---

## 🛠 Post-Deployment Tasks (Optional)


_Any migrations, scripts, or setup tasks?_

---

## 🎬 Fun GIF (Optional)


_What GIF best represents this PR?_


![fun gif](https://media.giphy.com/media/xT0xeJpnrWC4XWblEk/giphy.gif)

YassenAhmed-gif committed:
feat: Implement product filtering functionality

This commit introduces product filtering capabilities to the categories feature. Users can now sort products by various criteria like price, newness, and discount.

Key changes:
- Added `createdAt` field to `ProductEntity` and `ProductDtoMapper` to support sorting by newness.
- Implemented `discountPercent` getter in `ProductEntity` for consistent discount calculation.
- Introduced `FilterType` enum and a generic `Filter` class in `core/Functions/filter.dart` to handle sorting logic.
- Updated `CategoriesState` to include `filteredProducts` and `currentFilter`, and added `displayProducts` getter to show either filtered or all products.
- Modified `CategoriesViewModel` to:
    - Include `applyFilter` and `clearFilter` methods.
    - Initialize `filteredProducts` when fetching products.
- Added `FilterBottomSheet` widget to allow users to select filter options.
- Integrated `FilterButton` with `FilterBottomSheet` in `CategoryScreen`.
- Updated `CategoryScreenBody` to display `state.displayProducts` and show a "No products found" message when applicable.
- Modified `ProductGrid` to use the `discountPercent` from `ProductEntity`.
- Added new translation keys for filter options (e.g., "sort_by", "lowest_price", "new_product") in English and Arabic.
- Fixed a typo in an import statement in `main.dart` (`ececute_navigation` to `execute_navigation`).

YassenAhmed-gif committed:
Merge branch 'development' into feat/FECA-37-Filter

rajeh1032 commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/pages/category_screen.dart:
```diff
@@ -43,7 +44,16 @@ class CategoryScreen extends StatelessWidget {
+              return FilterButton(
+                currentFilter: state.currentFilter,
+                onFilterChanged: (filterType) {
+                  context.read<CategoriesViewModel>().applyFilter(filterType);
```
add event for applyFilter

rajeh1032 commented on lib/Feature/mainLayout/tabs/categoriesFeature/domain/entities/product_entity.dart:
```diff
@@ -15,8 +16,16 @@ class ProductEntity {
+      return ((1 - ((priceAfterDiscount ?? price!) / price!)) * 100).round();
+    }
+    return 0;
+  }
```
add this logic in use case

YassenAhmed-gif replied:
I need it in here becouse I get discountPercent in the filter function directly from the entity

rajeh1032 commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/viewModel/viewModel/categories_view_model.dart:
```diff
@@ -16,13 +18,65 @@ class CategoriesViewModel extends Cubit<CategoriesState> {
     this._getAllProductsUseCase,
   ) : super(const CategoriesState());
 
+  void applyFilter(FilterType? filterType) {
```
make this method private

ali-safwat commented on lib/core/Functions/filter.dart:
```diff
@@ -0,0 +1,49 @@
+enum FilterType {
```
it's prefered the enum values be all capital

rajeh1032 commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/viewModel/viewModel/categories_view_model.dart:
```diff
@@ -16,13 +18,65 @@ class CategoriesViewModel extends Cubit<CategoriesState> {
+    }
+  }
+
+  void clearFilter() {
```
make it private and add event if u use it in view

ali-safwat commented on lib/core/Functions/filter.dart:
```diff
@@ -0,0 +1,49 @@
+  highestPrice,
+}
+
+class Filter{
```
is this local filter

YassenAhmed-gif replied:
Yes

ali-safwat reviewed (changes requested):
we need to use api with query paramerters to filter the data not local filter it's good to use local filter but the main task is filter the api and query parameters

rajeh1032 commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/widgets/filter_bottom_sheet.dart:
```diff
@@ -0,0 +1,246 @@
+          children: [
+            Center(
+              child: Container(
+                width: 40,
```
use app size

rajeh1032 commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/widgets/filter_bottom_sheet.dart:
```diff
@@ -0,0 +1,246 @@
+          children: [
+            // Radio button
+            Container(
+              width: 20,
```
use app size

rajeh1032 commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/widgets/filter_bottom_sheet.dart:
```diff
@@ -0,0 +1,246 @@
+              child: isSelected
+                  ? Center(
+                child: Container(
+                  width: 8,
```
use app size

rajeh1032 commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/viewModel/states/categories_state.dart:
```diff
@@ -32,17 +39,28 @@ class CategoriesState extends Equatable {
+    }
+    return productsList ?? [];
+  }
+
```
make validation in view model

YassenAhmed-gif committed:
refactor: update filter functionality and event handling in categories feature

This commit refactors the filtering functionality within the categories feature.

Key changes include:
- Renamed `FilterType` enum to `FILTERTYPE` for consistency.
- Updated `CategoriesState` to use `FILTERTYPE` and removed the `displayProducts` getter, moving this logic to the `CategoriesViewModel`.
- Introduced `ApplyFilterEvent` and `ClearFilterEvent` in `CategoriesEvent` for better event management.
- Modified `CategoriesViewModel` to handle new filter events, updated filter application logic, and moved the `displayProducts` getter here.
- Adjusted `CategoryScreen`, `FilterButton`, and `FilterBottomSheet` to use the new `FILTERTYPE` and event handling.
- Updated `CategoryScreenBody` to use `context.watch<CategoriesViewModel>().displayProducts` and `state.isProductsLoading`.
- Replaced magic numbers with `AppSizes` constants in `FilterBottomSheet` and `FilterButton`.

YassenAhmed-gif committed:
Merge branch 'development' into feat/FECA-37-Filter
